torch/_refs/__init__.py

```diff
@@ -4278,20 +4273,79 @@ def arange(
 ) -> TensorLikeType:
     utils.check_layout(layout)
     utils.check_pin_memory(pin_memory)
+    device = torch.device(utils.device_or_default(device))
+
+    assert not isinstance(start, complex)
+    assert not isinstance(end, complex)
+    assert not isinstance(step, complex)
+
     # Case: torch.arange(5)
     if end is None:
         end = start
         start = 0
-    return prims.arange(
-        start,
-        end,
-        step,
-        dtype=dtype,
-        # layout=layout,
+    utils.check(step != 0, lambda: "step must be nonzero")
+    utils.check(
+        (step > 0 and end >= start) or (step < 0 and end <= start),
+        lambda: "upper bound and lower bound inconsistent with step sign",
+    )
+
+    def is_finite(x):
+        return not isinstance(x, FloatWithoutSymFloat) or math.isfinite(x)
+
+    utils.check(
+        is_finite(start) and is_finite(end),
+        lambda: f"unsupported range: {start} -> {end}",
+    )
+    utils.check(
+        is_finite(step),
+        lambda: f"step must be finite but got {step}",
+    )
+
+    if dtype is None:
+        args = (start, end, step)
+        integer_args = builtins.all(isinstance(arg, IntLike) for arg in args)
+        dtype = torch.int64 if integer_args else torch.get_default_dtype()
+
+    is_integer = utils.is_integer_dtype(dtype)
+    if is_integer:
+        xstart = sym_int(start)
+        xend = sym_int(end)
+        xstep = sym_int(step)
+
+    # For int64 we truncate arguments to int before calculating length, but
+    # other integral dtypes we don't. Weird... but needed to match ATen shapes.
+    if dtype == torch.int64:
+        length = math.ceil((xend - xstart) / xstep)
+    else:
+        print(start, end, step)
+        length = math.ceil((end - start) / step)
+
+    if is_integer:
+        return prims.iota(
+            length,
+            start=xstart,
+            step=xstep,
+            dtype=dtype,
+            device=device,
+            requires_grad=requires_grad,
+        )
+
+    computation_dtype = utils.get_acc_type(dtype, device)
+    index = prims.iota(
+        length,
+        start=0,
+        step=1,
+        dtype=torch.int64,
         device=device,
-        # pin_memory=pin_memory,
-        requires_grad=requires_grad,
+        requires_grad=False,
     )
+    index = _maybe_convert_to_dtype(index, computation_dtype)
+    result = start + step * index
```

Comment:
Is this alright, or is it better to do it via `lerp` for accuracy?

Reply:
This is the exact expression used in the `arange` kernels:
https://github.com/pytorch/pytorch/blob/3e6978172e86a97dc4c248406d2b2509a837299f/aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp#L33